### ISSUE TITLE
df = 0 in certain lme models

### ISSUE THREAD

**Issue by rvlenth** [rvlenth/emmeans#31]
User-reported issue:

In a model with 360 abundance measurement performed at six sampling dates, i want to know the pairwise tukey contrasts for the six dates.
```
mod <- glm(Abundance ~ Date, family="gaussian") 
```
FWIW, Abundance is normally distributed in its entirety, the subpopulations per date are not
```
means <- emmeans(mod, ~Date)
pairs (means, simple="Date")
```
works well, as does
```
glht(mod, mcp(Date="Tukey")).
```
However, `Abund` is spatially auto-correlated. I thus switch to `lme` with a dummy random effect and update to the correlation structure yielding the best AIC. 
```
best.mod <- lme(fixed=Abundance ~ Date, data = data, 
    random = ~1| dummy, method="ML", 
    correlation=corRat(1, form = ~ x + y))
```
Now, glht() still works:
```
> summary(glht(mods[[best.mod]], mcp(Date = "Tukey")))

	 Simultaneous Tests for General Linear Hypotheses

Multiple Comparisons of Means: Tukey Contrasts


Fit: lme.formula(fixed = Abundance ~ Date, data = D, random = ~1 | 
    Date, correlation = getFunction(m)(1, form = ~x + y), method = "ML")

Linear Hypotheses:
                  Estimate Std. Error z value Pr(>|z|)    
May - April == 0  -0.30829    0.15815  -1.949   0.3718    
June - April == 0  0.41849    0.15749   2.657   0.0841 .  
Aug. - April == 0 -0.25109    0.15741  -1.595   0.6018    
Oct. - April == 0 -0.28579    0.15747  -1.815   0.4561    
Nov. - April == 0 -0.09987    0.15709  -0.636   0.9883    
June - May == 0    0.72677    0.15759   4.612   <0.001 ***
Aug. - May == 0    0.05719    0.15751   0.363   0.9992    
Oct. - May == 0    0.02250    0.15758   0.143   1.0000    
Nov. - May == 0    0.20842    0.15720   1.326   0.7707    
Aug. - June == 0  -0.66958    0.15685  -4.269   <0.001 ***
Oct. - June == 0  -0.70428    0.15691  -4.488   <0.001 ***
Nov. - June == 0  -0.51835    0.15653  -3.311   0.0119 *  
Oct. - Aug. == 0  -0.03470    0.15683  -0.221   0.9999    
Nov. - Aug. == 0   0.15122    0.15645   0.967   0.9285    
Nov. - Oct. == 0   0.18592    0.15652   1.188   0.8429    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
(Adjusted p values reported -- single-step method)
```

However,  emmeans does not produce p-values with best.mod, possible because of zero degree of freedoms
```
> means <- emmeans(mods[[best.mod]], ~ Date)
> pairs(means, simple="Date")
 contrast        estimate        SE df t.ratio p.value
 April - May   0.30828775 0.1594914  0   1.933     NaN
 April - June -0.41848582 0.1588249  0  -2.635     NaN
 April - Aug.  0.25109352 0.1587444  0   1.582     NaN
 April - Oct.  0.28579102 0.1588097  0   1.800     NaN
 April - Nov.  0.09986911 0.1584278  0   0.630     NaN
 May - June   -0.72677357 0.1589291  0  -4.573     NaN
 May - Aug.   -0.05719423 0.1588486  0  -0.360     NaN
 May - Oct.   -0.02249673 0.1589139  0  -0.142     NaN
 May - Nov.   -0.20841864 0.1585322  0  -1.315     NaN
 June - Aug.   0.66957934 0.1581794  0   4.233     NaN
 June - Oct.   0.70427684 0.1582449  0   4.451     NaN
 June - Nov.   0.51835493 0.1578616  0   3.284     NaN
 Aug. - Oct.   0.03469750 0.1581641  0   0.219     NaN
 Aug. - Nov.  -0.15122440 0.1577806  0  -0.958     NaN
 Oct. - Nov.  -0.18592191 0.1578463  0  -1.178     NaN

P value adjustment: tukey method for comparing a family of 6 estimates 
Warning messages:
1: In pt(abs(t), DF, lower.tail = FALSE) : NaNs wurden erzeugt
2: In ptukey(sqrt(2) * abst, fam.size, zapsmall(df), lower.tail = FALSE) :
  NaNs wurden erzeugt
```

**Comment by rvlenth** [comment]
I note that the estimated diffferences are identical, but the SE’s are not. I also see that `glht()` uses z tests. If you do the `emmeans()` call with the additional argument `df = Inf`, I think it will at least give you results. 

In addition, if you add the argument `sigmaAdjust = FALSE`, I think the SEs will be the same.

If we get the SEs to match, then I think
```
pairs(mods[[best.mod]], adjust = "mvt")
```
will reproduce the P values you got from `glht()` (modulus the fact that they are obtained by a simulation method so there are slight variations in results from repeated calls).

**Comment by rvlenth** [comment]
If you have it handy still, could you please do:
```
    orig.means <- emmeans(mods[[best.mod]], ~ Date)
    orig.means@dfargs
    pairs(orig.lmeans)@linfct
```
and also
```
    mods[[best.mod]]$fixDF$X
```
... and send me the results?

**Comment by rvlenth** [comment]
I believe I have an explanation for this. I am guessing that the "dummy random effect" you used has but 1 level. If there is only one random effect, then there is only one group and that makes for zero d.f.

If this is what's happening, then the `@dfargs` will return `list(dfx = 0)`, or maybe a vector of zeros. This is arguably the right result when there is only one group. Why not use `gls()` instead of `lme()` to fit this model, given that you don't really need the random effect?

I think the proper remedy may indeed be specifying `df = `(whatever df you think is right) in the call. So I think I will leave everything as it stands unless I see that the situation is different than what I'm guessing.

**Comment by rvlenth** [comment]
I believe it's OK to close this issue now. It can always be re-opened if you see the need.

**Comment by hnguyen19** [comment]
I run into this issue after upgrading to version 1.7.1-1. I was contrasting the two levels of a factor and got p-value = `NaN` with the default `Tukey` adjustment. Switching to `mvt` gave me the p-value, but it is slightly different than the p-value I got for the same factor from `joint_test(model)`. Before updating, I was able to get the same p-values when I `contrast()` and `joint_test()`.

**Comment by rvlenth** [comment]
There are several points here. One is that the R function for Tukey's method does not work with very small degrees of freedom:
```
> ptukey(1.5, 2, df = c(1,1.5,2,3))
[1]       NaN       NaN 0.6000195 0.6333278
Warning message:
In ptukey(1.5, 2, c(1, 1.5, 2, 3)) : NaNs produced
```
... and that is likely where the NaN comes from.

Second, the `mvt` adjustment comes from a simulation technique, so the results are slightly different each time it is run. The same is true of the approximate Satterthwaite method used to obtain the d.f. from `lme` models.

Third, if you are really comparing just two means, then there is only one comparison, and so no multiplicity issue; the Tukey adjustment should be the same as the results with `adjust = "none"`. (But on the other hand, the code should automatically remove the Tukey adjustment in that case, so how is it that that didn't happen?)

I doubt this has anything to do with the new version. If it does, I would need a reproducible example in order to check what happens.

**Comment by hnguyen19** [comment]
Thank you very much for your quick response. Adding `adjust = none` did solve the problem. It doesn't look like the code automatically removes Tukey adjustment for me. Please find a reproducible example below.  

```
dput(oat)
structure(list(...1 = c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 
13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24), Plot = c(12, 
15, 21, 26, 31, 36, 43, 47, 11, 14, 23, 29, 32, 38, 48, 49, 15, 
18, 26, 28, 36, 37, 46, 47), Crop_ID = structure(c(2L, 1L, 2L, 
1L, 2L, 1L, 2L, 1L, 1L, 2L, 1L, 2L, 2L, 1L, 1L, 2L, 1L, 2L, 1L, 
2L, 1L, 2L, 2L, 2L), .Label = c("O3", "O4"), class = "factor"), 
    Rot = c("4-year", "3-year", "4-year", "3-year", "4-year", 
    "3-year", "4-year", "3-year", "3-year", "4-year", "3-year", 
    "4-year", "4-year", "3-year", "3-year", "4-year", "4-year", 
    "3-year", "3-year", "4-year", "3-year", "4-year", "4-year", 
    "4-year"), Variety = c("INO9201", "INO9201", "INO9201", "INO9201", 
    "INO9201", "INO9201", "INO9201", "INO9201", "INO9202", "INO9203", 
    "INO9204", "INO9205", "INO9206", "INO9207", "INO9208", "INO9209", 
    "INO9210", "INO9211", "INO9212", "INO9213", "INO9214", "INO9215", 
    "INO9216", "INO9217"), Plant = c("12-Apr-17", "12-Apr-17", 
    "12-Apr-17", "12-Apr-17", "12-Apr-17", "12-Apr-17", "12-Apr-17", 
    "12-Apr-17", "16-Apr-19", "16-Apr-19", "16-Apr-19", "16-Apr-19", 
    "16-Apr-19", "16-Apr-19", "16-Apr-19", "16-Apr-19", "2-Apr-20", 
    "2-Apr-20", "2-Apr-20", "2-Apr-20", "2-Apr-20", "2-Apr-20", 
    "2-Apr-20", "2-Apr-20"), Harvest = c("17-Jul-17", "17-Jul-17", 
    "17-Jul-17", "17-Jul-17", "17-Jul-17", "17-Jul-17", "17-Jul-17", 
    "17-Jul-17", "24-Jul-19", "24-Jul-19", "24-Jul-19", "24-Jul-19", 
    "24-Jul-19", "24-Jul-19", "24-Jul-19", "24-Jul-19", "24-Jul-20", 
    "24-Jul-20", "24-Jul-20", "24-Jul-20", "24-Jul-20", "24-Jul-20", 
    "24-Jul-20", "24-Jul-20"), Moisture = c(14.8, 14.5, 15.3, 
    15.3, 15.6, 14.8, 13.8, 13.5, 12.7, 13.4, 12.8, 11.9, 13.6, 
    15, 13.7, 11.1, 14.4, 15.3, 15.2, 15.35, 14, 14, 14, 14), 
    Standardized_yield_MgpHa = c(3.964908737, 3.764528097, 3.909064952, 
    3.439320172, 3.771097954, 3.248794318, 4.145579807, 2.848033037, 
    3.380191459, 3.633130956, 3.557577599, 3.416325673, 3.577287171, 
    3.396616101, 3.084547891, 2.096112896, 2.646739411, 3.27237047, 
    2.812118955, 2.987638692, 2.931228936, 3.823190701, 2.980556685, 
    2.670662357), Year = structure(c(1L, 1L, 1L, 1L, 1L, 1L, 
    1L, 1L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 3L, 3L, 3L, 3L, 3L, 
    3L, 3L, 3L), .Label = c("2017", "2019", "2020"), class = "factor"), 
    bu_p_acre = c("94.296875", "89.53125", "92.96875", "81.796875", 
    "89.6875", "77.265625", "98.59375", "67.734375", "80.390625", 
    "86.40625", "84.609375", "81.25", "85.078125", "80.78125", 
    "73.359375", "49.8515625", ".", ".", ".", ".", ".", ".", 
    ".", "."), lb_p_ac = c(3102.367188, 2945.578125, 3058.671875, 
    2691.117188, 2950.71875, 2542.039063, 3243.734375, 2228.460938, 
    2644.851563, 2842.765625, 2783.648438, 2673.125, 2799.070313, 
    2657.703125, 2413.523438, 1640.116406, 2070.957504, 2560.486368, 
    2200.359744, 2337.69624, 2293.55808, 2991.47904, 2332.15488, 
    2089.67616), Standardized_lb_ac = c(3536.698594, 3357.959063, 
    3486.885938, 3067.873594, 3363.819375, 2897.924531, 3697.857188, 
    2540.445469, 3015.130781, 3240.752813, 3173.359219, 3047.3625, 
    3190.940156, 3029.781563, 2751.416719, 1869.732703, 2360.891555, 
    2918.95446, 2508.410108, 2664.973714, 2614.656211, 3410.286106, 
    2658.656563, 2382.230822), Conversion_factor = c(892, 892, 
    892, 892, 892, 892, 892, 892, 892, 892, 892, 892, 892, 892, 
    892, 892, 892, 892, 892, 892, 892, 892, 892, 892), Standard_moisture = c(14, 
    14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 
    14, 14, 14, 14, 14, 14, 14, 14), Crop = c("oat grain", "oat grain", 
    "oat grain", "oat grain", "oat grain", "oat grain", "oat grain", 
    "oat grain", "oat grain", "oat grain", "oat grain", "oat grain", 
    "oat grain", "oat grain", "oat grain", "oat grain", "oat grain", 
    "oat grain", "oat grain", "oat grain", "oat grain", "oat grain", 
    "oat grain", "oat grain"), Block = structure(c(1L, 1L, 2L, 
    2L, 3L, 3L, 4L, 4L, 1L, 1L, 2L, 2L, 3L, 3L, 4L, 4L, 1L, 1L, 
    2L, 2L, 3L, 3L, 4L, 4L), .Label = c("rep1", "rep2", "rep3", 
    "rep4"), class = "factor"), block_trt = c("rep1/O4/NA", "rep1/O3/NA", 
    "rep2/O4/NA", "rep2/O3/NA", "rep3/O4/NA", "rep3/O3/NA", "rep4/O4/NA", 
    "rep4/O3/NA", "rep1/O3/NA", "rep1/O4/NA", "rep2/O3/NA", "rep2/O4/NA", 
    "rep3/O4/NA", "rep3/O3/NA", "rep4/O3/NA", "rep4/O4/NA", "rep1/O3/NA", 
    "rep1/O4/NA", "rep2/O3/NA", "rep2/O4/NA", "rep3/O3/NA", "rep3/O4/NA", 
    "rep4/O4/NA", "rep4/O4/NA"), Rot_n = c(4, 3, 4, 3, 4, 3, 
    4, 3, 3, 4, 3, 4, 4, 3, 3, 4, 4, 3, 3, 4, 3, 4, 4, 4)), row.names = c(NA, 
-24L), spec = structure(list(cols = list(...1 = structure(list(), class = c("collector_double", 
"collector")), Plot = structure(list(), class = c("collector_double", 
"collector")), Crop_ID = structure(list(), class = c("collector_character", 
"collector")), Rot = structure(list(), class = c("collector_character", 
"collector")), Variety = structure(list(), class = c("collector_character", 
"collector")), Plant = structure(list(), class = c("collector_character", 
"collector")), Harvest = structure(list(), class = c("collector_character", 
"collector")), Moisture = structure(list(), class = c("collector_double", 
"collector")), Standardized_yield_MgpHa = structure(list(), class = c("collector_double", 
"collector")), Year = structure(list(), class = c("collector_double", 
"collector")), bu_p_acre = structure(list(), class = c("collector_character", 
"collector")), lb_p_ac = structure(list(), class = c("collector_double", 
"collector")), Standardized_lb_ac = structure(list(), class = c("collector_double", 
"collector")), Conversion_factor = structure(list(), class = c("collector_double", 
"collector")), Standard_moisture = structure(list(), class = c("collector_double", 
"collector")), Crop = structure(list(), class = c("collector_character", 
"collector")), Block = structure(list(), class = c("collector_character", 
"collector")), block_trt = structure(list(), class = c("collector_character", 
"collector")), Rot_n = structure(list(), class = c("collector_double", 
"collector"))), default = structure(list(), class = c("collector_guess", 
"collector")), delim = ","), class = "col_spec"), problems = <pointer: 0x7fe716f788e0>, class = c("spec_tbl_df", 
"tbl_df", "tbl", "data.frame"))
```

model:
```
oat.lmer <- lmer(log(Standardized_yield_MgpHa) ~ Block + Crop_ID +  (1|Year) + (1|Year:Block) +  (1|Year:Crop_ID) + (1|Block:Year:Crop_ID),
  data=oat)
```
no adjustment, that matched `joint_test`'s p-value
```
joint_tests(oat.lmer)
 model term df1  df2 F.ratio p.value
 Block        3 5.56   2.859  0.1334
 Crop_ID      1 2.00   1.138  0.3979

contrast(emmeans(oat.lmer, ~ Crop_ID, infer = c(F,T),  type = "response"), method = "pairwise",  adjust = "none")
 contrast ratio     SE df null t.ratio p.value
 O3 / O4  0.913 0.0779  2    1  -1.067  0.3979

Results are averaged over the levels of: Block 
Degrees-of-freedom method: kenward-roger 
Tests are performed on the log scale 
```

no specification of adjustment and NaN p-value  
```
contrast(emmeans(oat.lmer, ~ Crop_ID, infer = c(F,T),  type = "response"), method = "pairwise")
 contrast ratio     SE df null t.ratio p.value
 O3 / O4  0.913 0.0779  2    1  -1.067     NaN

Results are averaged over the levels of: Block 
Degrees-of-freedom method: kenward-roger 
Tests are performed on the log scale 
Warning message:
In ptukey(sqrt(2) * abst, fam.size, zapsmall(df), lower.tail = FALSE) :
  NaNs produced
```

**Comment by rvlenth** [comment]
Hmmm, I am able to reproduce this, *after* modifying the source a bit. (I omitted everything after `row.names` and added `class = "data.frame"`).
```
> emmeans(oat.lmer, pairwise ~ Crop_ID, adjust = "tukey")
$emmeans
 Crop_ID emmean     SE   df lower.CL upper.CL
 O3        1.13 0.0744 3.71    0.919     1.35
 O4        1.22 0.0731 3.49    1.008     1.44

Results are averaged over the levels of: Block 
Degrees-of-freedom method: kenward-roger 
Results are given on the log (not the response) scale. 
Confidence level used: 0.95 

$contrasts
 contrast estimate     SE df t.ratio p.value
 O3 - O4   -0.0911 0.0853  2  -1.067     NaN

Results are averaged over the levels of: Block 
Degrees-of-freedom method: kenward-roger 
Results are given on the log (not the response) scale. 

Warning message:
In ptukey(sqrt(2) * abst, fam.size, zapsmall(df), lower.tail = FALSE) :
  NaNs produced
```
But the odd thing is this ain't the case with an `lm` model
```
> example(emmeans)
     ... output omitted ...

> emmeans(warp.lm, pairwise ~ wool, adjust = "tukey")
NOTE: Results may be misleading due to involvement in interactions
$emmeans
 wool emmean   SE df lower.CL upper.CL
 A      31.0 2.11 48     26.8     35.3
 B      25.3 2.11 48     21.0     29.5

Results are averaged over the levels of: tension 
Confidence level used: 0.95 

$contrasts
 contrast estimate   SE df t.ratio p.value
 A - B        5.78 2.98 48   1.940  0.0582

Results are averaged over the levels of: tension 
```
... where we indeed see that the Tukey adjustment was suppressed. Must be something very subtle. I don't see how this can depend on the model class, given that the adjustment-suppression code is in `summary.emmGrid`.

**Comment by rvlenth** [comment]
Seems to be related to degrees of freedom...
```
> update(TEST, df = 2)
 contrast estimate     SE df t.ratio p.value
 O3 - O4   -0.0911 0.0853  2  -1.067  0.3977

Results are averaged over the levels of: Block 
Degrees-of-freedom method: user-specified 
Results are given on the log (not the response) scale. 
```

**Comment by rvlenth** [comment]
OK, I have traced the problem to a logic error where I had recoded a complicated partial-patching expression to a custom internal partial-matching operator. That indeed is something new in the latest version. It seemed d.f.-related because the warning occurs with less than 2 d.f., and I could get the same warning by intercepting the summary of `WTEST` and making the df less than 2.

After fixing the logic error, here is a check that both CIs and tests pass without warnings:
```
> update(TEST, infer = TRUE)
 contrast estimate     SE df lower.CL upper.CL t.ratio p.value
 O3 - O4   -0.0911 0.0853  2   -0.459    0.277  -1.067  0.3979

Results are averaged over the levels of: Block 
Degrees-of-freedom method: kenward-roger 
Results are given on the log (not the response) scale. 
Confidence level used: 0.95 
```
Moreover, running in debug mode shows that we successfully remove the adjust method, causing the calls to `ptukey` and `qtukey` to never be reached.

Thanks for reporting this!
